File: test_currency_units.py
#!/usr/bin/env python3
"""Currency Unit Format Test — saves PNG screenshots to ./screenshots/"""

import asyncio
import os
import sys
from playwright.async_api import async_playwright

sys.stdout.reconfigure(encoding='utf-8', errors='replace')

SCREENSHOTS_DIR = "./screenshots"
os.makedirs(SCREENSHOTS_DIR, exist_ok=True)

BASE_URL = "http://76.13.17.91"
API_URL = "http://76.13.17.91:3001"


async def screenshot(page, path):
    full = os.path.join(SCREENSHOTS_DIR, path)
    await page.screenshot(path=full, full_page=True)
    print(f"  >> saved: {full}")
    return full


async def run():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        ctx = await browser.new_context(viewport={"width": 1440, "height": 900})
        page = await ctx.new_page()

        page.on("console", lambda msg: print(f"  [CONSOLE ERROR] {msg.text}") if msg.type == "error" else None)

        print("=== Currency Unit Format Test ===\n")

        # 0. API verify first
        print("[0] API unitType verification...")
        api_resp = await ctx.request.get(f"{API_URL}/api/datasets", timeout=10000)
        data = await api_resp.json()
        idr_datasets = []
        usd_datasets = []
        for ds in data["data"]:
            ut = ds.get("unitType", "MISSING")
            unit = ds.get("unit", "")
            label = f"  {ds['id'][:28]:28s} unitType={ut:15s} unit={unit}"
            print(label)
            if ut == "currency_idr":
                idr_datasets.append(ds)
            elif ut == "currency_usd":
                usd_datasets.append(ds)

        print(f"\n  IDR datasets: {[d['id'] for d in idr_datasets]}")
        print(f"  USD datasets: {[d['id'] for d in usd_datasets]}")

        # 1. Homepage
        print("\n[1] Homepage...")
        await page.goto(f"{BASE_URL}", wait_until="domcontentloaded", timeout=30000)
        await page.wait_for_timeout(3000)
        await screenshot(page, "01_homepage.png")

        # 2. DataHub list page
        print("\n[2] Data Hub list page...")
        await page.goto(f"{BASE_URL}/data/market-dashboard", wait_until="domcontentloaded", timeout=30000)
        await page.wait_for_timeout(3000)
        await screenshot(page, "02_datahub_market.png")

        # 3. Navigate to each IDR dataset page
        print("\n[3] IDR dataset detail pages...")
        for i, ds in enumerate(idr_datasets[:3]):
            try:
                await page.goto(f"{BASE_URL}/data/{ds['id']}", wait_until="domcontentloaded", timeout=20000)
                await page.wait_for_timeout(2000)
                await screenshot(page, f"03_idr_{i+1}_{ds['id']}.png")
                print(f"  IDR {ds['id']} saved")
            except Exception as e:
                print(f"  ERR IDR {ds['id']}: {e}")

        # 4. Navigate to each USD dataset page
        print("\n[4] USD dataset detail pages...")
        for i, ds in enumerate(usd_datasets[:3]):
            try:
                await page.goto(f"{BASE_URL}/data/{ds['id']}", wait_until="domcontentloaded", timeout=20000)
                await page.wait_for_timeout(2000)
                await screenshot(page, f"04_usd_{i+1}_{ds['id']}.png")
                print(f"  USD {ds['id']} saved")
            except Exception as e:
                print(f"  ERR USD {ds['id']}: {e}")

        # 5. Admin panel
        print("\n[5] Admin panel — unit type editor...")
        await page.goto(f"{BASE_URL}/admin", wait_until="domcontentloaded", timeout=30000)
        await page.wait_for_timeout(3000)
        await screenshot(page, "05_admin.png")

        # Try to find a dataset row and click it
        try:
            rows = await page.query_selector_all("tr, [class*='border']")
            print(f"  Found {len(rows)} rows/elements in admin table")
            for row in rows[:5]:
                txt = await row.inner_text()
                if len(txt) > 10:
                    print(f"    row: {txt[:60]}")
        except Exception as e:
            print(f"  Admin row scan: {e}")

        print("\n=== COMPLETE ===")
        print(f"Screenshots: {SCREENSHOTS_DIR}/")
        await browser.close()


if __name__ == "__main__":
    asyncio.run(run())

File: test-currency-units.js
const { chromium } = require('@playwright/test');

async function run() {
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    viewport: { width: 1440, height: 900 }
  });

  const page = await context.newPage();

  // Enable console logging
  page.on('console', msg => {
    if (msg.type() === 'error') console.log('  [CONSOLE ERROR]', msg.text());
  });

  console.log('=== Currency Unit Format Test ===\n');

  try {
    // 1. Go to VPS homepage first
    console.log('[1] Navigating to homepage...');
    await page.goto('http://76.13.17.91', { waitUntil: 'networkidle', timeout: 30000 });
    console.log('  ✓ Homepage loaded:', await page.title());
    await page.screenshot({ path: 'screenshots/01_homepage.png' });
    console.log('  → saved: screenshots/01_homepage.png');

    // 2. Navigate to Data Hub
    console.log('\n[2] Navigating to Data Hub...');
    await page.goto('http://76.13.17.91/data', { waitUntil: 'networkidle', timeout: 30000 });
    await page.screenshot({ path: 'screenshots/02_datahub.png' });
    console.log('  → saved: screenshots/02_datahub.png');

    // 3. Get all dataset cards and their unit values
    console.log('\n[3] Scanning datasets on Data Hub...');
    const datasetCards = await page.$$eval('[class*="border"][class*="cursor-pointer"]', els =>
      els.map(el => {
        const title = el.querySelector('[class*="font-semibold"]')?.textContent?.trim() || '';
        const value = el.querySelector('[class*="text-2xl"], [class*="text-2"]')?.textContent?.trim() || '';
        const unit = el.querySelector('[class*="text-gray-400"]')?.textContent?.trim() || '';
        return { title, value, unit };
      })
    );
    console.log('  Found cards:', datasetCards.length);

    // 4. Find IDR and USD datasets
    console.log('\n[4] Looking for IDR and USD datasets...');
    const idrSets = datasetCards.filter(d => d.unit?.includes('IDR'));
    const usdSets = datasetCards.filter(d =>
      d.unit?.includes('USD') || d.value?.includes('USD') || d.value?.includes('$')
    );

    console.log('  IDR datasets:', idrSets.length, idrSets.map(d => d.title));
    console.log('  USD datasets:', usdSets.length, usdSets.map(d => d.title));

    // 5. Navigate to each IDR dataset detail
    if (idrSets.length > 0) {
      for (const ds of idrSets) {
        console.log(`\n[5] Testing IDR dataset: ${ds.title}`);
        console.log(`    Displayed value: "${ds.value}"`);
        console.log(`    Unit label: "${ds.unit}"`);

        // Click the card to open detail view
        const card = await page.$(`text="${ds.title}"`);
        if (card) {
          await card.click();
          await page.waitForTimeout(2000);
          await page.screenshot({
            path: `screenshots/03_idr_${ds.title.replace(/\s+/g, '_').replace(/[^\w]/g, '')}.png`
          });
          console.log(`  → saved screenshot`);

          // Go back to datahub
          await page.goBack();
          await page.waitForTimeout(1000);
        }
      }
    }

    // 6. Navigate to each USD dataset
    if (usdSets.length > 0) {
      for (const ds of usdSets) {
        console.log(`\n[6] Testing USD dataset: ${ds.title}`);
        console.log(`    Displayed value: "${ds.value}"`);
        console.log(`    Unit label: "${ds.unit}"`);

        const card = await page.$(`text="${ds.title}"`);
        if (card) {
          await card.click();
          await page.waitForTimeout(2000);
          await page.screenshot({
            path: `screenshots/04_usd_${ds.title.replace(/\s+/g, '_').replace(/[^\w]/g, '')}.png`
          });
          console.log(`  → saved screenshot`);
          await page.goBack();
          await page.waitForTimeout(1000);
        }
      }
    }

    // 7. Test via Admin panel if available
    console.log('\n[7] Testing Admin panel — unit type editor...');
    await page.goto('http://76.13.17.91/admin', { waitUntil: 'networkidle', timeout: 30000 });
    await page.screenshot({ path: 'screenshots/05_admin.png' });
    console.log('  → saved: screenshots/05_admin.png');

    // Click on a dataset to open editor
    const datasetItems = await page.$$('[class*="cursor-pointer"], [class*="clickable"]');
    if (datasetItems.length > 0) {
      await datasetItems[0].click();
      await page.waitForTimeout(1500);
      await page.screenshot({ path: 'screenshots/06_admin_editor.png' });
      console.log('  → saved: screenshots/06_admin_editor.png');

      // Check for unit type selector
      const unitTypeSelect = await page.$('select');
      if (unitTypeSelect) {
        const options = await unitTypeSelect.$$eval('option', opts => opts.map(o => o.textContent));
        console.log('  Unit type options found:', options);
      }
    }

    // 8. Final summary
    console.log('\n=== Test Summary ===');
    console.log('All screenshots saved to ./screenshots/');
    console.log('IDR datasets tested:', idrSets.map(d => d.title).join(', '));
    console.log('USD datasets tested:', usdSets.map(d => d.title).join(', '));
    console.log('Status: COMPLETE');

  } catch (err) {
    console.error('\n❌ Test error:', err.message);
    await page.screenshot({ path: 'screenshots/error_state.png' });
    console.log('Error screenshot saved: screenshots/error_state.png');
  } finally {
    await browser.close();
  }
}

// Ensure screenshots dir exists
const fs = require('fs');
if (!fs.existsSync('screenshots')) fs.mkdirSync('screenshots');

run();
